apps/bfd-model/bfd-model-rif/src/main/resources/db/scripts/BFD-2572-Cleanup-Medicare-Beneficairyid-History/invoke-psql.sh

```diff
@@ -0,0 +1,329 @@
+        AND (B.EFCTV_END_DT IS NULL
+                  OR B.EFCTV_END_DT = A.MBI_EFCTV_END_DT)
+UNION
+-- matches between beneficiaries_history and medicare_beneficiaryid_history_temp will be deleted
+--
+SELECT A.BENE_MBI_ID
+      FROM MEDICARE_BENEFICIARYID_HISTORY_TEMP A,
+        BENEFICIARIES_HISTORY B
+      WHERE B.BENE_ID = A.BENE_ID
+        AND B.MBI_NUM = A.MBI_NUM
+        AND B.EFCTV_BGN_DT = A.MBI_EFCTV_BGN_DT
+        AND (B.EFCTV_END_DT IS NULL
+                  OR B.EFCTV_END_DT = A.MBI_EFCTV_END_DT) );
+EOM
+
+# STEP_FOUR_CLEANUP_SQL
+# ==============================================================
+read -r -d '' STEP_FOUR_CLEANUP_SQL << EOM
+DELETE
+FROM MEDICARE_BENEFICIARYID_HISTORY_TEMP D
+WHERE D.BENE_MBI_ID in (
+   SELECT V1.BENE_MBI_ID
+   FROM BENEFICIARIES_HISTORY C,
+   (
+      -- find records in medicare_beneficiaryid_history that match
+      -- on bene_id but do not match on mbi_num. it simply means that
+      -- the beneficiary's current mbi_num is not equal to any of
+      -- the mbi_num(s) in the medicare_beneficiaryid_history table.
+      -- This will potentially be an 'add' if there is no matching data
+      -- already in the beneficiaries_history table.
+      --
+      SELECT A.BENE_MBI_ID,
+             A.BENE_ID,
+             A.MBI_NUM,
+             A.MBI_EFCTV_BGN_DT AS EFCTV_BGN_DT,
+             A.MBI_EFCTV_END_DT AS EFCTV_END_DT
+      FROM MEDICARE_BENEFICIARYID_HISTORY_TEMP A,
+           BENEFICIARIES B
+      WHERE B.BENE_ID = A.BENE_ID
+      AND B.MBI_NUM <> A.MBI_NUM
+   ) V1
+-- find medicare_beneficiaryid_history_temp records that
+-- match beneficiaries_history data already present.
+WHERE (C.BENE_ID = V1.BENE_ID
+AND C.MBI_NUM = V1.MBI_NUM));
+EOM
+
+# STEP_FIVE_INSERT_SQL
+# ==============================================================
+# SQL to extract data from the medicare_beneficiaryid_history_temp
+# and beneficiares tables and create records in the beneficiaries_history
+# table.
+#
+# we want any records that are in the medicare_beneficiaryid_history
+# BUT ARE NOT in the beneficiaries_history table. Also, we'll grab
+# a bene's sex, birth date and HICN info from our current beneficiaries
+# table to have a somewhat more complete set of data for those records
+# inserted into the beneficiaries_history table. We'll use now() for
+# last_updated in case we need to back out the INSERTs.
+# ==============================================================
+read -r -d '' STEP_FIVE_INSERT_SQL << EOM
+INSERT INTO BENEFICIARIES_HISTORY (
+  BENE_HISTORY_ID,
+  BENE_ID,
+  MBI_NUM,
+  EFCTV_BGN_DT,
+  EFCTV_END_DT,
+  BENE_CRNT_HIC_NUM,
+  HICN_UNHASHED,
+  BENE_SEX_IDENT_CD,
+  BENE_BIRTH_DT,
+  LAST_UPDATED)
+SELECT
+  NEXTVAL('beneficiaryhistory_beneficiaryhistoryid_seq'),
+  V2.BENE_ID,
+  V2.MBI_NUM,
+  V2.EFCTV_BGN_DT,
+  V2.EFCTV_END_DT,
+  X.BENE_CRNT_HIC_NUM,
+  X.HICN_UNHASHED,
+  X.BENE_SEX_IDENT_CD,
+  X.BENE_BIRTH_DT,
+  NOW()
```

Comment:
Is this going to snapshot the same for all the entires, or these all going to be slightly offset based on how long it takes to do the inserts?

Reply:
they will be slightly offset; however, we are talking a total job processing time of < 1h, so we will know pretty quickly if we need to back things out and would be done well before any subsequent weekend processing. Creating all same timestamp(s) to meis not reflective of when the update **really** happened.

Reply:
Just thinking more along the lines if they need to be backed out, but it'd still be possible.